### ISSUE TITLE
Asynchronously update model Design Doc

### ISSUE THREAD

**Issue by chunyang-wen** [sql-machine-learning/elasticdl#1095]
In distributed model training, there may be more than one workers. Each of them can calculate gradients and report gradients to parameter servers. Parameter servers can choose from following two ways of updating gradients upon receiving updating request:
+ Synchronously
  + Parameter servers make sure that all the workers pull the same version of parameters. And they do not apply the updates until they receive all workers' gradients. After updating, all workers start to pull parameters and repeat.
+ Asynchronously
  + Parameter servers start to apply gradients update when any of the workers sends its gradients to parameter servers.

Synchronously updating model usually generates better models. It can make sure that all workers are synchronised and none of the parameters is stale when updating happens. But its performance can deteriorate when slow worker appears.

Asynchronously updating model has a better training speed. But usually it takes a longer time to converge comparing to synchronously updating model.

In ElasticDL, we have implemented synchronously updating model. We acquire a lock when we receive gradient from workers and pre-aggregate them. When all the workers have reported the gradient, we update model and model version.

In order to support asynchronously updating model, we need to change updating function into a stateless function. This function receives gradients and updates model.

**TODO** Describe detailed changes in order to support asynchronously updating model.

+ [Revisiting distributed synchronous SGD](https://arxiv.org/pdf/1604.00981.pdf)
+ [Synchronous_Stochastic_Gradient](https://www.cs.ubc.ca/labs/lci/mlrg/slides/MLRG_Synchronous_Stochastic_Gradient.pdf)
+ [tf.train.SyncReplicaOptimizer](https://www.tensorflow.org/api_docs/python/tf/train/SyncReplicasOptimizer)
